File: problems/leetcode/0100/155/155.cpp
// 155. Min Stack
// https://leetcode.com/problems/min-stack/description/

#include <utility>
#include <vector>

using namespace std;

typedef pair<int, int> pii;

class MinStack {
public:
  vector<pii> valueAndMinimums;

  MinStack() {}

  void push(int val) {
    if (valueAndMinimums.size() == 0) {
      valueAndMinimums.push_back({val, val});
    } else {
      valueAndMinimums.push_back(
          {val, min(val, valueAndMinimums.back().second)});
    }
  }

  void pop() { valueAndMinimums.pop_back(); }

  int top() { return valueAndMinimums.back().first; }

  int getMin() { return valueAndMinimums.back().second; }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: problems/leetcode/0100/155/155.py
# 155. Min Stack
# https://leetcode.com/problems/min-stack/description/

class MinStack:

    def __init__(self):
        self.valAndCumulativeMins = []

    def push(self, val: int) -> None:
        cumulativeMin = val
        if len(self.valAndCumulativeMins) > 0:
            cumulativeMin = min(cumulativeMin, self.valAndCumulativeMins[-1][1])
        self.valAndCumulativeMins.append((val, cumulativeMin))

    def pop(self) -> None:
        self.valAndCumulativeMins.pop()

    def top(self) -> int:
        return self.valAndCumulativeMins[-1][0]

    def getMin(self) -> int:
        return self.valAndCumulativeMins[-1][1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
